### PR TITLE
reporter api

### PR DESCRIPTION
**Pull request by lucamrgs**
Defines the initial API endpoints and route logic for reporting

### CODE REVIEW

File: test/unittest/mocks/mock_reporter/mock_downstream_reporter.go [hunk 1]
```diff
@@ -28,3 +30,13 @@ func (reporter *Mock_Downstream_Reporter) ReportStepEnd(executionId uuid.UUID, s
...
+
+func (reporter *Mock_Downstream_Reporter) GetExecutionsIds() []string {
+	args := reporter.Called()
+	return args.Get(0).([]string)
+}
+
+func (reporter *Mock_Downstream_Reporter) GetExecutionReport(executionKey uuid.UUID) (cache_report.ExecutionEntry, error) {
+	args := reporter.Called(executionKey)
+	return args.Get(0).(cache_report.ExecutionEntry), args.Error(1)
+}
```
Comment (MaartendeKruijf) [inline test/unittest/mocks/mock_reporter/mock_downstream_reporter.go:42]: These should be in a separate mock as they are not in the cache interface

File: test/unittest/reporters/downstream_reporter/cache_test.go [hunk 2]
```diff
@@ -4,7 +4,7 @@ import (
 	"errors"
 	"soarca/internal/reporter/downstream_reporter/cache"
 	"soarca/models/cacao"
-	"soarca/models/report"
+	report "soarca/models/cache"
```
Comment (MaartendeKruijf) [inline test/unittest/reporters/downstream_reporter/cache_test.go:7]: Here is a mismatch between models and naming

File: test/integration/api/reporter_api_test.go
Comment (MaartendeKruijf) [inline test/integration/api/reporter_api_test.go:1]: Place this in the unittest folder as it's an unittest

File: docs/content/en/docs/core-components/api-reporter.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+|status             |execution-status-enum  |string             |The current [status](#execution-stataus) of the execution
+|status_text        |explanation            |string             |A natural language explanation of the current status or related info
+|error              |error                  |string     |Error raised along the execution of the playbook at execution level
+|step_results       |step_results           |dictionary         |Map of step-id to related [step execution data](#step-execution-data)
+|request_interval   |seconds                |integer            |Suggests the polling interval for the next request (default suggested is 5 seconds).
+
+
+##### Step execution data
+|field              |content                |type               | description |
+| ----------------- | --------------------- | ----------------- | ----------- |
```
Comment (MaartendeKruijf) [inline docs/content/en/docs/core-components/api-reporter.md:81]: Automated execution field is missing here

File: docs/content/en/docs/core-components/api-reporter.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+##### Execution stataus 
+Table from [Cyentific RNI workflow Status](https://github.com/cyentific-rni/workflow-status/blob/main/README.md#21-refined-execution-status-enumeration)
+**Vocabulary Name:** `execution-status-enum`
+| Property Name | Description|
+| :--- | :--- |
+| successfully_executed | The workflow step was executed successfully (completed). |
+|failed| The workflow step failed. |
+|ongoing| The workflow step is in progress. |
+|server_side_error| A server-side error occurred. |
```
Comment (MaartendeKruijf) [inline docs/content/en/docs/core-components/api-reporter.md:99]: I've sugested await_user_input https://github.com/cyentific-rni/workflow-status/pull/3 if that goes true we need to add that here

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]
Nice start on the api we need to refine a bit but looks good already

File: docs/content/en/docs/core-components/api-reporter.md (hunk 3 above)
Comment (MaartendeKruijf) [reply docs/content/en/docs/core-components/api-reporter.md:81]: Also see what other fields would make sense to include

**Review by MaartendeKruijf — CHANGES_REQUESTED** [review]

File: docs/content/en/docs/core-components/api-reporter.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+|status_text        |explanation            |string             |A natural language explanation of the current status or related info
+|error              |error                  |string     |Error raised along the execution of the playbook at execution level
+|step_results       |step_results           |dictionary         |Map of step-id to related [step execution data](#step-execution-data)
+|request_interval   |seconds                |integer            |Suggests the polling interval for the next request (default suggested is 5 seconds).
+
+
+##### Step execution data
+|field              |content                |type               | description |
+| ----------------- | --------------------- | ----------------- | ----------- |
+|execution_id       |UUID                   |string             |The id of the execution of the playbook where the step resides
```
[the remark this answers is not in the document]
Comment (lucamrgs) [reply docs/content/en/docs/core-components/api-reporter.md:82]: Took me forever but now it's done

File: routes/reporter/reporter_api.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	executionEntryParsed, err := parseCachePlaybookEntry(executionEntry)
+	if err != nil {
+		log.Debug("Could not parse entry to reporter result model")
+		SendErrorResponse(g, http.StatusInternalServerError, "Could not parse execution report", "GET /report/{id}")
+		return
+	}
+	g.JSON(http.StatusOK, executionEntryParsed)
+}
+
+func SendErrorResponse(g *gin.Context, status int, message string, orginal_call string) {
```
[the remark this answers is not in the document]
Comment (lucamrgs) [reply routes/reporter/reporter_api.go:70]: What do you mean?

File: routes/reporter/reporter_api.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+// getExecutions GET handler for obtaining all the executions that can be retrieved.
+// Returns this to the gin context as a list if execution IDs in json format
+//
+//	@Summary	gets all the UUIDs for the executions that can be retireved
+//	@Schemes
+//	@Description	return all stored executions
+//	@Tags			reporter
+//	@Produce		json
+//	@success		200	{array}	string
+//	@error			400
```
[the remark this answers is not in the document]
Comment (lucamrgs) [reply routes/reporter/reporter_api.go:31]: You mean @param?

[the remark this answers is not in the document]
Comment (lucamrgs) [reply routes/reporter/reporter_api.go:31]: Do you mean the "@ param" comment?

File: test/unittest/routes/reporter_api/reporter_api_test.go [hunk 8]
```diff
@@ -103,13 +103,13 @@ func TestGetExecutions(t *testing.T) {
...
-			StepResults:    map[string]cache_model.StepResult{},
-			PlaybookResult: nil,
-			Status:         2,
+			ExecutionId: executionId,
+			PlaybookId:  "test",
+			Started:     expectedStarted,
+			Ended:       expectedEnded,
+			StepResults: map[string]cache_model.StepResult{},
+			Error:       nil,
+			Status:      2,
```
Comment (MaartendeKruijf) [inline test/unittest/routes/reporter_api/reporter_api_test.go:112]: Status 2 is an enum right?

**Review by MaartendeKruijf — DISMISSED** [review]

**Review by MaartendeKruijf — APPROVED** [review]

**Pull request merged**